assembl/static2/js/app/components/common/treeItem.jsx

```diff
@@ -0,0 +1,267 @@
+/* eslint react/no-multi-comp: "off" */
+// @flow
+import * as React from 'react';
+import { Map } from 'immutable';
+import classnames from 'classnames';
+import debounce from 'lodash/debounce';
+import { CellMeasurerCache, List } from 'react-virtualized';
+
+import { scrollToPost } from '../../utils/hashLinkScroll';
+import NuggetsManager from './nuggetsManager';
+
+type BaseProps = {
+  id: string,
+  identifier: string,
+  phaseId: string,
+  level: number,
+  fullLevel?: string,
+  rowIndex: number,
+  hidden: boolean,
+  originalLocale?: string,
+  lang: string,
+  contentLocaleMapping: Map<string, string>,
+  listRef: List,
+  nuggetsManager: NuggetsManager,
+  cache: CellMeasurerCache,
+  SeparatorComponent: React.Node,
+  InnerComponentFolded: ({ nbPosts: number }) => React.Node,
+  InnerComponent: (
+    BaseProps & {
+      contentLocale: string,
+      numChildren: number,
+      measureTreeHeight: (delay?: number) => void
+    }
+  ) => React.Node
+};
+
+type Props = {
+  children: Array<TreeItem>
+} & BaseProps;
+
+type State = {
+  expanded: boolean,
+  visible: boolean
+};
+
+class Child extends React.PureComponent<Props, State> {
+  static defaultProps = {
+    InnerComponentFolded: () => null,
+    level: 0,
+    hidden: false
+  };
+
+  state = {
+    expanded: true,
+    visible: false
+  };
+
+  componentDidMount() {
+    this.onScroll();
+    window.addEventListener('scroll', this.onScroll);
+    window.addEventListener('resize', this.onScroll);
+  }
+
+  componentWillUnmount() {
+    this.stopListening();
+  }
+
+  holder: HTMLDivElement | null = null;
+
+  scrollAnchor: { current: null | HTMLDivElement } = React.createRef();
+
+  onScroll = debounce(() => {
+    const holder = this.holder;
+    if (!holder) {
+      return;
+    }
+    const box = holder.getBoundingClientRect();
+    const pageYOffset = window.pageYOffset;
+    const top = box.top + pageYOffset;
+    // visible if the top of the box is in viewport or next page
+    const isVisible = top < pageYOffset + 2 * window.innerHeight && top > pageYOffset;
```

Comment:
Just wondering why this calculation with "2 * window.innerHeight"

Reply:
one window.innerHeight in the viewport, two is the viewport of the next page